StepContainerV2: Rewrite for composition and DX

zaguiini opened the pull request:
## Proposed Changes

From the title, it rewrites the internals of `StepContainerV2`, aiming for extensibility, composition, and ease of maintenance.

This is a big PR, and I found it difficult to split it into chunks, so here's a commit-by-commit summary with the most relevant changes and their 'why.'

### Rewrite StepContainerV2 for composition

Before, `StepContainerV2` accepted the slots and juggled them according to some props, like `verticalAlign`.

Now, `StepContainerV2` returns a `div` with important metadata about the step being rendered, such as the `is[Small|Large]Viewport` info and the height of the current top and sticky bottom bars.

Because of this change, each wireframe defines what, how, and where each slot will be rendered. This might lead to some repetition between wireframes, but it's for the greater good. This kind of composition ensures we won't end up with complexity down the road while keeping the API surface area small enough to be easy to understand and extend.

This separation also allows us to better place functionality that should be limited to specific wireframes, but are outside of the content area, which is the only place customizable "from the outside" for any wireframe. For example, it didn't make sense to have the `hasContentPadding` prop exposed for all wireframes when it only makes sense to expose this setting to the `FullWidth` layout. This doesn't happen anymore :)

### Fix vertical alignment and component measurements

This was a huge problem, but not anymore!

I was struggling a lot with measuring the top and sticky bottom bars for components that need to be changed (example: design picker preview, although it doesn't currently use it). This happened because we needed to read the dimensions of these elements, but since we only had a monolith (`StepContainerV2`) and were rendering that component on the server, this failed.

Now, we're only allowing to vertically center the wireframes that could be vertically-centered, like `CenteredColumnLayout`. This is done by a helper component called `<TopBarRenderer />`, which is called within the wireframe and wraps the `topBar` slot from the step.

"But it uses `useLayoutEffect` under the hood, which doesn't work on the server."

Yes, you're right. The loading step is unique in various ways: it's rendered on the server, and it has vertically-centered content even on mobile. Because of it, I think it warranted its wireframe. It makes a lot of sense, and it simplifies things a lot. Specifically, this wireframe does not use the `<TopBarRenderer />`. It renders `<TopBar />` as it is (we don't want it customized anyway) and calls it a day. 👨🏻‍🍳 💋

### Rename NextButton to PrimaryButton and add SecondaryButton too
### Do not use the label prop to define the content of the button
### Rename StickyBottomBar props
### Rename TopBar props

These are imported from a previous draft PR and are mostly naming changes. It makes more sense to use these names for precision. You do not always do want to render a `skip button` in `TopBar` - we're rendering the help center in a few steps.

### Remove Heading.maxWidth prop

This isn't needed as we're no longer stretching the content wrapper horizontally (unless using the FullWidth component of course, but in this specific case you want the contents to stretch.)

## Testing Instructions

There should be no visual and functional regressions compared to production. You should be able to go side by side between versions and see no changes.

zaguiini committed:
Rename NextButton to PrimaryButton and add SecondaryButton too

zaguiini committed:
Rewrite StepContainerV2 for composition

zaguiini committed:
Fix vertical alignment and component measurements

zaguiini committed:
Remove unnecessary loading wrapper component

zaguiini committed:
Do not use the label prop to define the content of the button

zaguiini committed:
Fix stories

zaguiini committed:
Rename StickyBottomBar props

zaguiini committed:
Rename TopBar props

zaguiini committed:
Remove Heading.maxWidth prop

zaguiini committed:
Update README

vykes-mac commented:
I did some test comparing both desktop and mobile side by side (prod on the right). The videos might be a little longer than usual. Everything looks pretty good.

https://github.com/user-attachments/assets/1eb9c676-5732-429f-b2aa-a434c08c4d53


https://github.com/user-attachments/assets/4d1804ce-6d88-4f0d-8b8d-865d360f580e

At **2:22** in the mobile video, before navigating away from checkout, theres a small frame `You have no item in your cart` message component flash. That does not happen in the production verision.
![image](https://github.com/user-attachments/assets/05dd4232-c2c2-43be-a13f-1b3085d32c67)

p-jackson committed:
Fix up eslint errors

p-jackson commented:
> Because of this change, each wireframe defines what, how, and where each slot will be rendered. This might lead to some repetition between wireframes, but it's for the greater good. This kind of composition ensures we won't end up with complexity down the road while keeping the API surface area small enough to be easy to understand and extend.

💯 I already liked this idea, but now that I see it all written out, I'm even more convinced. [Duplication is better than the wrong abstraction](https://sandimetz.com/blog/2016/1/20/the-wrong-abstraction), and there's no design/product reason why the different wireframes should be re-using the same `div`, so we shouldn't try and abstract over the coincidental code duplication.

>  it didn't make sense to have the hasContentPadding prop exposed for all wireframes when it only makes sense to expose this setting to the FullWidth layout

❤️

autumnfjeld commented:
@luis  We should have testing instructions on all PRs.

@tellthemachines   You can follow these instructions:

* Use the calypso live link
 * Test the step at /setup/onboarding
  * Ensure you also test using ?flags=-onboarding/step-container-v2 too to confirm that when the flag isn't

Or test in local dev with  /setup/onboarding

p-jackson committed:
Remove @ts-expect-error comment

Explicitly casting to the expected CSSProperties type is safer.
@ts-expect-error will ignore _all_ errors, even if they're typos. The
explicit cast is a much narrower set of things that get ignored.

p-jackson committed:
Removed children prop from StickyBottomBar

No one uses this. Let's not provide a children slot if we're not sure
why anyone would use it.

p-jackson commented on client/my-sites/marketing/do-it-for-me/faq.tsx:
```diff
@@ -18,6 +18,7 @@ const FAQHeader = styled.h1`
 `;
 
 const FAQSection = styled.div`
+	width: 100%;
```
@zaguiini this rule doesn't seem to have any affect 🤔 
I tested the FAQ section on 
`/start/do-it-for-me/new-or-existing-site`, and
`/setup/site-setup/difmStartingPoint?siteSlug={{ site }}`

And neither cared whether this `width: 100%` was there 🤷

p-jackson replied:
I'm not going to remove it though because you may have need to add it :)

p-jackson reviewed (approved):
Lots of changes to review, but mostly renames, so not too bad.
This encapsulates a lot of the changes we've been chatting about over the last week. Very cool to see them all come together.

I pushed a couple small tweaks, but mostly this was good as it 👍

tellthemachines commented:
I've given this a few runs on desktop and mobile and can't find anything out of place. I can hardly spot any changes except for the improved header button positioning:

in this PR
<img width="430" alt="Screenshot 2025-03-31 at 2 15 01 pm" src="https://github.com/user-attachments/assets/28b68faa-bd77-4c65-a23c-e9ec6ea522d6" />

vs in trunk
<img width="425" alt="Screenshot 2025-03-31 at 2 14 50 pm" src="https://github.com/user-attachments/assets/4d3f627b-34d3-45c1-9371-57c3e8e155c3" />

Looks good in testing!
